### PR TITLE
Quarry 3.0

### LINKED ISSUE #147: Support for alternate HV Quarry dig pattern
Hello.  The current HV Quarry digs "in front" of the quarry engine.  I'd like to have a version that digs below the quarry, with the quarry at the center of the "square radius" that is dug out.  Eg, if the quarry is at "(x, y, z)", then the volume dug is from "(x - r, y - 1, z - r)" to "(x + r, y - 101, z + r)".  Also, add digiline support for stopping, starting, restarting and getting a status report (current depth).

If I added this functionality, would a PR be accepted and merged?  I have a few options:

1) Create an entirely new HV quarry node type (technic:hv_quarry_2 or similar), and allowing crafting between the two quarry engines to convert one to the other type (similar to how a lua controller <-> microcontroller craft works).
2) Augment the existing HV quarry, so that the formspec has options for specifying the offset from center of hole from the 3d node position of the quarry itself.

In any event, I would obviously not change the behavior of the existing quarry engine in such a way that would break or alter any existing uses of it.

Thoughts?

### PR DESCRIPTION
**Pull request by OgelGames**
Another feature and performance rewrite for the quarry.

**Main changes:**
- Added configurable offsets (x/z) to the quarry's digging area.
- The quarry now only digs below itself, to allow for the new offsets.
- Quarries no longer have a rotation, and have been changed to match visually.
- The quarry formspec is now updated on right-click, instead of on every run.
- Less metadata usage.
- Improved working area visualization (see below).

Doing some basic testing by timing the running of 20 quarries, I found an improvement of **45%** when running, and **68%** while idle. The majority of the saving is from not updating the formspec every run, but there are also less calculations done when running.

The new visualization uses particles spawners to show the edges of the working area:

Implements and closes #147 (ping to @dennisjenkins75)

### CODE REVIEW

**Comment by BuckarooBanzay** [comment]
Nice! I like the use of particlespawners for the region (and might steal this idea on another project :wink:)

I had my test-setup of 120+ quarries on the pandorabox test-server and decided to update the `technic` mod to this branch and see if/how it behaves.

Sadly every quarry i already placed was rotated with "face" down and had a param2 of `2` (not sure if that's relevant)

screenshot_20221031_131440

Oh, and somehow the front-texture was missing/blacked-out :/

**Comment by OgelGames** [comment]
@BuckarooBanzay that looks correct, quarries don't have a rotation anymore, so I changed the design to look like there is a hole on the bottom. (I forgot to mention that sorry)

**Comment by S-S-X** [comment]
> The quarry now only digs below itself, to allow for the new offsets.

Debatable but I'd like this to be configurable because it apparently prevents using forward (and a bit of upward too) digging quarry ships I've used a lot.

**Comment by OgelGames** [comment]
> Will the old quarry behavior be supported as well (so that existing ship designed that assume the old behavior will continue to function)?

Yes, existing quarries will be configured by an LBM to match their previous digging area, so they will continue to work the same. The only exception is changing the radius, as that doesn't behave the same way as before.

> > The quarry now only digs below itself, to allow for the new offsets.
> 
> Debatable but I'd like this to be configurable because it apparently prevents using forward (and a bit of upward too) digging quarry ships I've used a lot.

It's not possible to dig up anymore, otherwise the quarry would dig itself and whatever is around it.

**Comment by S-S-X** [comment]
> > > The quarry now only digs below itself, to allow for the new offsets.
> > 
> > 
> > Debatable but I'd like this to be configurable because it apparently prevents using forward (and a bit of upward too) digging quarry ships I've used a lot.
> 
> It's not possible to dig up anymore, otherwise the quarry would dig itself and whatever is around it.

I can add functionality for that if needed, would be just complaining if selected offsets overlap quarry itself.

Another way which could bring both compatibility with old methods and new content could be introducing Quarry Mk2 where old quarry could simply use exact same code but with limited configuration and old tiles?

But I guess only if someone else would also see use for old functionality or introducing another tier with different functionality...

**Comment by OgelGames** [comment]
> Another way which could bring both compatibility with old methods and new content could be introducing Quarry Mk2 where old quarry could simply use exact same code but with limited configuration and old tiles?

I really want to avoid adding another version of the quarry, not only because the old version will be mostly obsolete, but also communicating about them will get confusing.

> But I guess only if someone else would also see use for old functionality

So far only FeXoR and MCLV have commented on the new quarry (see #offtopic), and FeXoR doesn't mind:
> [Libera] \<FeXoR\> Ah, right, just digging up won't be possible any more. A minor downside IMO ;)

Need to get some more players testing them.

**Review by BuckarooBanzay — APPROVED** [review]

**Comment by BuckarooBanzay** [comment]
vletter

nice :+1: 

One thing though: could it be possible that a switching station begins to run in one place and trigger the `technic_run` of an not-yet converted (old) quarry somewhere else? (not anywhere near the player to trigger the migration lbm)
If so: what would happen then?

**Comment by S-S-X** [comment]
> I tested it and LBMs do require players to trigger them

LBM is named after Load but it is actually executed on Activate instead of block loading...

**Comment by OgelGames** [comment]
> LBM is named after Load but it is actually executed on Activate instead of block loading...

Yep... I should have just looked at the docs: https://github.com/minetest/minetest/blob/88820cd31c2fe21c8c16ee547a3335131e8ba009/doc/lua_api.txt#L7891-L7893

I added a simple check to prevent quarries digging before they've been updated, they wouldn't crash but they would start digging with the default offsets of zero.

File: technic/machines/HV/quarry.lua
Comment (S-S-X) [inline technic/machines/HV/quarry.lua:269]: I think this shouldn't really be configurable per quarry but instead be global server configuration.

It is more environment effect for all players, not something that quarry owner should control.
If configurable per quarry then there probably should be another server configuration that enables or disables per quarry configuration...

My opinion: better to leave this to be server wide configuration and remove per quarry config + meta values so that effects as seen by players are same no matter what quarry node is removing nodes near them.

File: technic/machines/HV/quarry.lua
Comment (S-S-X) [inline technic/machines/HV/quarry.lua:371]: Could reduce digiline action time a bit by using bit better performing find instead of sub, LuaJIT does not optimize this case. Potentially reduces penalty a bit when mesecons_debug is enabled and many quarries connected to network.
I've not tested actual penalty but it is not only performance, find is also shorter and reduces potential index errors (you can see immediately if it is correct or not without counting characters or thinking about indexing method).

So faster, cleaner and reduces human error surface:
```suggestion
		elseif msg:find("^radius ") then
```

File: technic/machines/HV/quarry.lua
Comment (OgelGames) [reply technic/machines/HV/quarry.lua:269]: Hmm, you make a good point... I think I will remove that.

File: technic/machines/HV/quarry.lua
Comment (OgelGames) [reply technic/machines/HV/quarry.lua:371]: Interesting, I would have assumed that getting part of a string would be faster than finding a pattern 🤔

Comment (OgelGames) [reply technic/machines/HV/quarry.lua:371]: Did a quick test running code in-game (not sure how accurate it is), comparing the two methods for 1,000,000 calls:

- `string.sub` with LuaJIT on: 337μs
- `string.find` with LuaJIT on: 71795μs
- `string.sub` with LuaJIT off: 20788μs
- `string.find` with LuaJIT off: 70662μs

Unless I'm doing something wrong with testing, it seems like my assumption is correct 🤔

Comment (S-S-X) [reply technic/machines/HV/quarry.lua:363]: Yes, you're almost sure doing it wrong and JIT is optimizing out test code for you removing most of it. You need to use dynamic input to prevent caching, input caching isn't possible in actual use cases.
To prevent optimizing out code you can add counter for input string and even better make every other case fail in test loop with alternating input data.

Comment (S-S-X) [reply technic/machines/HV/quarry.lua:363]: Also tested with actual minetest engine and found out that performance was actually very different from plain luajit run and with mt string.sub clearly won, very interesting thing was that sub was nearly 10x faster with all mods enabled but only 2x with mods disabled.
Did quick search for overrides as that would immediately explain it but didn't find any.

What you got seems right with mt engine and sub indeed performs better and probably should be left as is (your JIT results still seems skewed which is probably optimization).

Comment (OgelGames) [reply technic/machines/HV/quarry.lua:363]: Did another test with this code, trying to make the input dynamic a bit:
```lua
local t = minetest.get_us_time()
for i=1, 100000 do
    local s = string.rep(i % 2 == 1 and "a" or "b", (i % 10) + 10)

    local x = s:sub(1, 5)
    -- OR
    local x = s:find("^aaaaa")
end
print(minetest.get_us_time() - t)
```
Subtracting the time it takes to run the loop without either `sub` or `find`, the times are much closer, but still `sub` wins.

Either way, we're dealing with microseconds of difference when it comes to the actual use case, so it really doesn't matter which one is used.

**Comment by OgelGames** [comment]
Did some more testing, specifically for the previous bug in #143, and it stops correctly at `max_depth`.

I did find a weird thing where the quarry instantly finds the first node to dig (it should take several cycles), so I'll have to figure that out before I merge this.

**Comment by OgelGames** [comment]
Turns out the ground search was using absolute positions, so it was finding the ground on the first run, and the digging loop was finding the ground (I guess it must be well optimized :D).

Also added a depth check to the ground search, and loading of the area.

File: technic/machines/HV/quarry.lua
Comment (S-S-X) [reply technic/machines/HV/quarry.lua:378]: Yeah it seems sub is faster anyway and true that this really does not need optimization for performance by itself.

Actually started discussion only for mesecons_debug penalty mechanism because it is a lot more sensitive (nothing wrong with that) and small things can make difference there: even just few % faster already allows commanding more devices.

> Subtracting the time it takes to run the loop without either sub or find, the times are much closer

Remember that code which creates input string is same for both and increases execution time, for accurate performance measurement. After getting results you'd also have to subtract time used up by utility stuff or it will skew proportions.

**Comment by OgelGames** [comment]
I think this is ready to merge now, it's at least safe enough for live testing :boom:

**Pull request merged**
